perfkitbenchmarker/publisher.py

```diff
@@ -655,37 +658,56 @@ class SampleCollector(object):
     samples: A list of Sample objects.
     metadata_providers: A list of MetadataProvider objects. Metadata providers
       to use.  Defaults to DEFAULT_METADATA_PROVIDERS.
-    publishers: A list of SamplePublisher objects. If not specified, defaults to
-      a LogPublisher, PrettyPrintStreamPublisher, NewlineDelimitedJSONPublisher,
-      a BigQueryPublisher if FLAGS.bigquery_table is specified, and a
-      CloudStoragePublisher if FLAGS.cloud_storage_bucket is specified. See
-      SampleCollector._DefaultPublishers.
+    publishers: A list of SamplePublisher objects to publish to.
+    publishers_from_flags: If True, construct publishers based on FLAGS and add
+      those to the publishers list.
+    add_default_publishers: If True, add a LogPublisher,
+      PrettyPrintStreamPublisher, and NewlineDelimitedJSONPublisher targeting
+      the run directory to the publishers list.
     run_uri: A unique tag for the run.
   """
-  def __init__(self, metadata_providers=None, publishers=None):
+  def __init__(self, metadata_providers=None, publishers=None,
+               publishers_from_flags=True, add_default_publishers=True):
     self.samples = []
 
     if metadata_providers is not None:
       self.metadata_providers = metadata_providers
     else:
       self.metadata_providers = DEFAULT_METADATA_PROVIDERS
 
-    if publishers is not None:
-      self.publishers = publishers
-    else:
-      self.publishers = SampleCollector._DefaultPublishers()
+    self.publishers = publishers.copy()
```

Comment:
Ah shoot, I forgot to run the tests. Thank you! I wonder why this error didn't trigger when I ran the code.

I copy the list because we're modifying it. If someone passed in a list and then later reused that list for something else, they might be surprised to find the extra elements in it. I admit it's an unlikely scenario here, but I try to keep my functions clean because bugs like that can be really difficult to track down.